### DISCUSSION TITLE
Should every curb polygon require a unique ID?

### DISCUSSION THREAD

**Discussion by jacobbaskin** [openmobilityfoundation/curb-data-specification#2] [Ideas]
Hi all,

Since yesterday's meeting, another implementation question I've been thinking about has been unique IDs for curb space. It's pretty clear to me that any curb space that has event information needs a unique identifier, but we will have lots of curb space without events. So the question there is, should we require in the spec that every polygon of curb space with its own regs info gets its own ID.

The benefits of having IDs is obvious, but there are some challenges that are worth considering:

  * IDs may encourage people to think of them as stable when they really aren't. For instance, if I auto-generate IDs based on a linear reference, those IDs can change due to very slight updates to the underlying data (base map changes, segment splits/joins, etc)
  * Conversely, stable IDs can encourage people to apply regulations anachronistically -- e.g. I look at the current regs and I see that space 12345 is 40 feet long and allows loading only, but a month ago that space was 50 feet long and also allowed parking, so if I have an event stream from a month ago I will misinterpret it.
  * Requiring IDs Imposes additional requirements on data producers
  * A single conceptual stretch of curb can be broken up into multiple regulations if there are e.g. driveways or buffers separating spaces, so the space you want to give an ID to may not match the space you want to regulate
  * If we use CurbLR-style regulation priorities, different polygons can overlap each other in space, so assigning each one a unique ID means that a point in space could correspond to multiple IDs.
 
Would appreciate thoughts on this!

**Comment by schnuerle** [comment]
I think it makes sense and is pretty easy to create a unique ID for each curb space polygon. Whether it gets used or not, it's worth doing it for when it will be used. And I can see the ID being used in MDS to reference CDS curb areas.

For the format, I recommend using [UUID](https://github.com/openmobilityfoundation/mobility-data-specification/blob/main/general-information.md#uuids) (Universally Unique Identifiers), which is an easily generated unique and random identifier. We use UUIDs extensively in MDS and they work well.

**Reply by jacobbaskin** [reply]
So the last bullet point is the hard one here. Let's say I have a curb with little 2-foot red-painted buffers in between every other metered parking stall (very common in California). What gets an ID? Is it:
- The whole row of parking stalls?
- Every uniquely-regulated segment (so, every group of 2 spots PLUS every 2-foot stretch of red paint)?
- Only the stalls and not the buffers (if so, how does this work)?

**Reply by cartograham** [reply]
@jacobbaskin : this is James Graham from DDOT.  Question for you - is it the physical curb space that you are proposing should have unique identifiers?  Or is it the regulations which are present _within_ that physical curb space?  In the DDOT world, we uniquely ID each blockface/side and reference it against our LRS.  This ID is relatively static and will not change unless there is a significant street reconfiguration.  

I am also in favor of having a dynamic/potentially changing unique ID for the curb regs that fall within the physical curb space - as long as that same regulation maintains a link/reference to that physical blockface.  Like @schnuerle said - it's an easy thing to implement and I don't think it's overly burdensome on cities.

**Reply by schnuerle** [reply]
I think overlapping polygons in space is fine. If there is a hierarchy of areas like Space -> Curb Area -> Area of Interest. I don't think there should be buffer polygons built into the spec. I think you define the polygon areas as accurately as you can, define how hierarchy works, have some guidance on overlap (eg, don't overlap spaces, but you can overlap areas of interest), and maybe provide guidance on buffers for GPS in practice. But I think there will be self and automatic reporting of curb area selection going on in the field, some GPS snapping maybe, and GPS cannot be the only method for knowing if something is in or out of an area.

**Reply by jacobbaskin** [reply]
So here are two examples from Seattle that illustrate what I'm trying to say:

**Case 1**
![Screen Shot 2021-06-04 at 2 28 11 PM](https://user-images.githubusercontent.com/20517826/120847178-1f15ca80-c541-11eb-9c51-28ab531851da.png)

Here, we can see that the right-hand stretch of curb has two regulations during the day but only one regulation overnight. So the question is, do we represent the overnight regulation as a single segment (with its own ID) or as two different segments with the same regulations?

**Case 2**
![Screen Shot 2021-06-04 at 2 26 15 PM](https://user-images.githubusercontent.com/20517826/120847287-47052e00-c541-11eb-9774-91ba6d03468e.png)

Here we see the same regulation (paid parking) across a curb but with little gaps interspersed -- these are the "buffers" I'm referring to. The question is, do these gaps get their own IDs too?

**Reply by schnuerle** [reply]
In case 1 I think we have 2 segments with same regs.

In case 2 I don't think the gaps get any IDs.  

IMO, I think we need to use CDS solve for access to areas that are "commercial" loading curb zones, and have an eye in the spec for future uses around other activity. I'm not sure the best way to do this until we get into the spec design and start to look at gaps in the spec structure and heirarchy.

**Reply by jacobbaskin** [reply]
Yeah makes sense! I think those are good answers to start with -- the case 1 answer is what we do at Coord but doesn't align with CurbLR which is why I was pushing on it.

**Reply by jacobbaskin** [reply]
I am going to propose having unique IDs be provider-defined and recommending UUIDs rather than requiring UUIDs. The reason for this is that UUIDs aren't generally computed from existing data, and we for instance will want to dynamically compute unique IDs for curb spaces based on their linear reference data rather than assigning them in advance.

**Reply by schnuerle** [reply]
In that case maybe both IDs are good to require.  UUIDs are good as a randomly created, unique identifier that every defined area can get. A more simple ID could be used as well, derived from existing data. Do you have an example of this second ID? Are you thinking some sort of psuedo readable ID with info in it, like a SKU? There would also a third type of ID, a way to reference external IDs.

**Reply by jacobbaskin** [reply]
The problem with requiring a UUID is that you have to store it somewhere, which I would prefer not to do if we don't have to. Usually when we want to create a unique ID, we do something like hash together a list of uniquely identifying information -- so in this case we might do like SHA256(curb segment ID + distance in meters rounded to nearest 0.1)

**Reply by schnuerle** [reply]
Where would the curb segment ID come from, the city's base street segment data, or underlying LR data?  

What if the paint or curb cut or sign gets changed, and you add another foot to the available curb distance - do you need a new ID for that same curb segment?

In MDS the UUIDs have worked well to identify and cross link things as needed. Yes they sometimes have to be stored somewhere (file or database). Some are dynamic and created on the fly, or they are generated once and never change (like [provider IDs](https://github.com/openmobilityfoundation/mobility-data-specification/blob/main/providers.csv)).

**Reply by jacobbaskin** [reply]
It would be up to the provider. I just don't see why having a particular ID be UUID-formatted needs to be a thing; it's just a string anyway, so who cares?

**Reply by jacobbaskin** [reply]
We will need to write up some stuff about when you can expect IDs to be stable. For pure regulation data, to a first approximation I would say "never". But if you define space as being monitored, bookable, etc. then probably we will say something like the ID will remain stable as long as the use of the space stays substantially the same.

**Reply by schnuerle** [reply]
Since the ID is then not strictly dependent on the underlying data, then I think UUID is good mostly because of consistency with MDS.  SHA256 (eg, e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) is a string not much different that a UUID (eg, 123e4567-e89b-12d3-a456-426655440000).

**Reply by jacobbaskin** [reply]
Yeah I had an idea in my head that UUIDs had some internal structure that made them hard to just stick hashes into but it looks like generating UUIDv5s from hashes is in fact encouraged, so let's use 'em

**Reply by schnuerle** [reply]
Great. Yeah the hashes are there to simply provide some structure to the string, and v5 could work well if seed data is needed. In other cases I imagine no seed data is needed, so could use the random UUID v4.

**Reply by jacobmalleau** [reply]
To follow up James' point, I like the idea of having unique IDs for segment blocks (segIDs) while regulations have their own unique IDs (regIDs) like how you have described above. I know this brings in the question of how you label the segment blocks - but many cities working at the curbside have some type of identification for this.

As for the two cases you provided, I agree with the approach to Case 1, however I don't love the idea of leaving gaps in the curb to be dealt with later. At CurbIQ we have tried to emphasize mapping the entire curb and found that cities don't just want to know where the 'commercial loading zones' are, but what all curbside space is designated as (curb bulbs, no parking, etc.). By leaving those gaps, it then becomes complicated when updating and changing designations. Say that whole street becomes No Parking - you would then have No Parking segments with gaps in between with the proposed approach, opposed to having the whole curbside be No Parking (including the small 2' gaps segments) if you were to have all segments with unique IDs.

**Comment by jiffyclub** [comment]
I think every entity should have a unique ID, so the question for me is what are the atomic entities in CDS? In CurbLR the atomic entity seems to be a stretch of curb and its associated regulations (if any). I think that should also include dates of validity.

**Reply by schnuerle** [reply]
I think in CDS the atomic entity is Curb Zone (a polygonal area of curb that is used for some specific activities), which aligns to CurbLR in purpose I believe. There is also optionally a level lower in CDS, which is Curb Space (a designated place for a vehicle).

**Comment by jacobbaskin** [comment]
I think this has been pretty resolved in favor; locking.
